### ISSUE TITLE
Create threat model

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#177]
into Wiki
Kinda standard, but https://github.com/PrivateBin/PrivateBin/issues/174 shows we need it…

Maybe boring doc stuff and many (at least for us) obvious things would be added, but I think it is a good thing to do.
I think we need to separate the different options though (with/without password, with/without burn after reading, …)

It would also standardize the "What this does/does not" section in the Readme…

**Comment by rugk** [comment]
Here is a good example of a threat model: https://github.com/google/end-to-end/wiki/Threat-model

**Comment by hasufell** [comment]
You can also check more semi-formal methods like https://download.microsoft.com/download/8/1/6/816C597A-5592-4867-A0A6-A0181703CD59/Microsoft_Press_eBook_TheSecurityDevelopmentLifecycle_PDF.pdf under Chapter 9, Stage 4 "Risk Analysis" (page 101).

Another method is to grind the [MITRE CAPEC database](https://capec.mitre.org/) for attack patterns that may apply, then organize them in conjunction with assets/architecture and propose mitigations.

**Comment by elrido** [comment]
If you guys want one, please feel free to create one. I think our current docs quite clearly document this already. That is maybe what is @rugk means by "(at least for us) obvious things"?

**Comment by elrido** [comment]
I'll give this a try with my basic common sense assumptions and structure as per the example. Will definitely require review and extension.

**Comment by rugk** [comment]
Oh already looks great.

> neither the administrator, the server or a third party changes the JavaScript of PrivateBin

I would not take this into the assumptions part. As this, which cannot be just assured, is the reason why we e.g. neeed HTTPS etc. Same with "the user can trust the admin and the server operators".
I would rather explain it as a consequence.
E.g. by listing example attacks, which are (outside of our) threat model. Okay, giving the meaning of "outside threat model" I think it is actually a good idea to list it as a "assumption". So forget my first part. But anyway, listing example attacks to explain why the assumptions are needed would be a good idea.
Also maybe list attacks, which are explicitly included, i.e. also list things you want to include in your thread model. (may be data extraction/XSS, decryption, mainly in our case, I think)

**Comment by rugk** [comment]
I would also drop that "the cryptographic libraries are used correctly". We cannot assure this and this claim is bold. Actually this is also in our threat model, because we want to be made aware of, if they are not used correctly.

**Comment by rugk** [comment]
> Assumptions: Maybe semantics, but I understood the document that I should list even implicit assumptions that may often be incorrect in reality. 

Just had a small look at the PDF; but mhh… can you quote an exact page?
In any case, I'd still say that assumptions, should include the implicit ones, which we know are sure. But assumptions are things you "expect"/"take for given". ANd we cannot take our own implementation for given. Just leave that out.

Also you don't have to follow that guide to strictly, it's just one (kinda outdated – I really like how they say storing encryption keys in some whatever thing in WInXP is good, while in Linux it is not :laughing: ) book…
Also you certainly don't have top do that numeric stuff at the end/evaluating how high a "change of attack" is. We are not a company that may need these numbers to decide whether they want to fix a bug or not. :wink: 

So under "Define Use Scenarios" is also says you should state the _reasons_ why one thing might not apply.

**Comment by elrido** [comment]
Disclaimer: I have quite a lot of experience with risk assessments, but none with threat models.

You provided one example of an output (which does not tell me how to do it) and, thankfully, provided a PDF which explains the process of creating one in great detail. I did invest two evenings reading that book, then sat down for one evening taking notes and doing sketches according to pages 101-114. Finally I spent another evening drawing the diagrams on the computer and writing down properly my notes and ordering them into the Wiki page.

As far as I understand from the book, the goal is to create a risk assessment to be able to decide what may need fixing/improving. To that end we need those numbers, because they are the rational basis of what has the highest priority. What we have now is just the input of the model, the actual modeling is still missing as far as I understand. My understanding was that we need to do this, so that we can rationally decide what risks we may have in this project that we may or may not be aware of and also that we can explain to others why we do focus on certain things when they send in PRs or suggest certain features (say the "send an email when something gets commented" one).

Assumptions: Page 106, the Linux thing you quoted is exactly such an example of an assumption that may not be true on all environments. So it would have been noted in that model and subsequently the Linux example would have been used to disproof it, hence creating a risk (in the risk part that is still missing in ours so far). From the [OD](https://en.oxforddictionaries.com/definition/assumption): "A thing that is accepted as true or as certain to happen, without proof." Most of these statements are quite certainly not correct. We know that some users don't trust the admin, we know that in many countries there is that kind of network surveillance, we know that if an admin creates a paste on his own, the key and content is known to the admin. So these are (flawed) assumptions that need to go into that model, so that they pop out as risks. We certainly agree that correct implementation this is not necessarily true. I hope you agree, that we should try to achieve this and improve it if we find an incorrect use. But it remains our responsibility IMHO.

The Win/Linux thing actually makes sense: In Win, you would use the OS capability for (cryptographically) protected storage, not plain files stored readable to anyone with admin privs or physical access. In Linux there is no such built in mechanism and your app has to handle that by itself. IMHO I wouldn't trust Win there to do this correct and it doesn't protect you from implementing that facility just as wrong as the crypto library you would use to build it in yourself.

Now, either you take it over from here, since you actually seem to know much better about what you want to have in this document then me, or you let me complete the process and then we can discuss about the risk assessments, etc. Or maybe you would like me to use a different process to create such a document? Or would like us to do it using a shared etherpad or similar?

What do you want / how should we proceed? I am a bit frustrated at the moment after all this work and you basically telling me this process is wrong, when it was you who instructed me to do it as in the PDF.

**Comment by rugk** [comment]
Well... I also have no idea how thread modelling works. 😄 

Just saw these examples and the PDF, just like you. I just don't want to exaggerate this whole thing. When I created this issue, I did not have this 100%ly formal approach in mind, but just to create some threat model, i.e. what are our threats/what are in scope of the project/threat model, which are not, etc. (And BTW, it was not me who provided the PDF.)
That said, of course, I have nothing against this formal approach and if it helps the project - then it's totally fine.

> We know that some users don't trust the admin, we know that in many countries there is that kind of network surveillance, we know that if an admin creates a paste on his own, the key and content is known to the admin. So these are (flawed) assumptions that need to go into that model, so that they pop out as risks. We certainly agree that correct implementation this is not necessarily true. I hope you agree, that we should try to achieve this and improve it if we find an incorrect use.

Yeah, that's it! That's the difference I mean. Correct code = Never true, this is what we want to spot erros on. No network advisary = Well... also not true, but we don't care, because we say "Use HTTPS". Thus, we may leave it out. However, considering metadata leaks and such stuff, we of course care about a MITM actor.
That's why this thing is maybe a bit too formal. Or at least I don't understand whether to include that thing as an assumption or not.
To give a typical assumption, you included, which we have to make e.g.: "Scjl is without [security] flaws.". That's also certainly not true, but not in our scope, as we cannot fix it there.
Or same thing: "Admin is 100%ly trusted." Here, given our design, we cannot protect against such an attacker. Thus outside of thread model.

>  In Win, you would use the OS capability for (cryptographically) protected storage

Which is [certainly not secure](https://github.com/gentilkiwi/mimikatz). So as you say, you may not trust Win (XP in the example!) to do it correctly. As such, it is also the dev's responsibility to evaluate the libs/features you choose of course and not trust 100%ly blindly. But I guess, we agree here.

----

So end results: No I did not instruct you to this PDF, I have not even read it before you presented me with this huge result, and yes, I actually do like it and think it is not wrong at all. I was just impressed by this hard work you put into it (so props to you 💪 ). As said, I actually just thought of a much smaller and lighter task (maybe a file of the size of the Readme, maybe), so I just want to prevent you from doing too much work. But if you like that, or maybe you find a vulnerability by that (😉) please continue.
I only had some small remarks about some points, which I just think I would e.g. not consider a problem or not. My approach here is not this formal one, but just saying what is in threat model/what not - similar to how Bug Bounties or so have these definitions. All I thought this issue is about is just clearly documenting, what we anyway knew for some time, i..e stating the "for us obvios", and maybe extending it with things, one has to think about. Generally I just thought "Let's think about what threats we could have and write them down. And the same with ones, we explicitly do not have." If this is the end result (and it seems so from skimming the PDF) then this is totally fine. It's just more complicated (and likely better, systematic, thoroughgoing) formal approach than I would have choosen.
I have nothing against if you finish it. If you have experience in risk assessment (and threat modelling) you have certainly more experience than me, so go on. I'll stay quiet until this is done. 😉 The end result is what counts and as far as I understand (and you explained) the current stuff is just preparation.
I am sorry if it looked like I would not like what you've done here so far. It's actually great. Just so you know...

**Comment by elrido** [comment]
Ok, will work on the release as discussed, tomorrow, then will get back to this. I could maybe mark obviously flawed assumptions to avoid misinterpretation.

I don't think that HTTPS is the solution, end-to-end encryption and the web of trust is AFAIK the better one. State actors have access to their own certificate authorities, can generate valid certs on the fly and Germany for example (FinFisher) is known to sell software for ISPs to inject malware into otherwise benign HTTP(S) data streams. So certainly better then nothing, but not a panacea.

**Comment by rugk** [comment]
No need to discuss HTTPS here. We know it's flaws, but we know it's the only thing we have. In browsers, at least. And we know PrivateBin is only secure with HTTPS. That's just the status quo, which we cannot change as a web application.
Web of trust has never succeeded in the way PKI has.

**Comment by elrido** [comment]
Updated the doc with the threat assessment table and suggested mitigations for the key risks. I also added some sections clarifying the intent of the "assumptions" and that this focuses on the server side, not client (so not much about crypto in here).

We will have to extend this further down the line, but I leave it as a basis for that future work, for now. My personal key take-aways are:
1. We need the paste manager / paste federation to make UI and server more independent.
2. While somewhat obvious, software needs to stay compatible with up to date environments (say, PHP or browsers) so an ongoing mitigation is the maintenance of the software. Shows the value of testing our software on so many PHP versions and not just a single one.
